deployment/data-feeds/changeset/solana/deploy_cache.go

```diff
@@ -0,0 +1,496 @@
+	cldf "github.com/smartcontractkit/chainlink-deployments-framework/deployment"
+	"github.com/smartcontractkit/chainlink-deployments-framework/operations"
+	"github.com/smartcontractkit/chainlink/deployment/common/proposalutils"
+	seq "github.com/smartcontractkit/chainlink/deployment/data-feeds/changeset/solana/sequence"
+	"github.com/smartcontractkit/chainlink/deployment/data-feeds/changeset/solana/sequence/operation"
+	"github.com/smartcontractkit/chainlink/deployment/helpers"
+
+	df_cache "github.com/smartcontractkit/chainlink-solana/contracts/generated/data_feeds_cache"
+)
+
+const (
+	CacheContract datastore.ContractType = "DataFeedsCache"
+	CacheState    datastore.ContractType = "DataFeedsCacheState"
+)
+
+type DeployCacheRequest struct {
+	ChainSel    uint64
+	BuildConfig *helpers.BuildSolanaConfig
+	Qualifier   string
+	LabelSet    datastore.LabelSet
+	Version     string
+	FeedAdmins  []solana.PublicKey // Feed admins to be added to the cache
+}
+
+var _ cldf.ChangeSetV2[*DeployCacheRequest] = DeployCache{}
+
+type DeployCache struct{}
+
+func (cs DeployCache) VerifyPreconditions(env cldf.Environment, req *DeployCacheRequest) error {
+	if _, ok := env.BlockChains.SolanaChains()[req.ChainSel]; !ok {
+		return fmt.Errorf("solana chain not found for chain selector %d", req.ChainSel)
+	}
+	if _, err := semver.NewVersion(req.Version); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (cs DeployCache) Apply(env cldf.Environment, req *DeployCacheRequest) (cldf.ChangesetOutput, error) {
+	var out cldf.ChangesetOutput
+
+	if req.BuildConfig != nil {
+		// You may want to define a specific build params for the cache if needed
+		err := helpers.BuildSolana(env, *req.BuildConfig, cacheBuildParams)
+		if err != nil {
+			return out, fmt.Errorf("failed build solana artifacts: %w", err)
+		}
+	}
+
+	out.DataStore = datastore.NewMemoryDataStore()
+	version := semver.MustParse(req.Version)
+	ch, ok := env.BlockChains.SolanaChains()[req.ChainSel]
+	if !ok {
+		return out, fmt.Errorf("solana chain not found for chain selector %d", req.ChainSel)
+	}
+
+	deploySeqInput := seq.DeployCacheSeqInput{
+		ChainSel:    req.ChainSel,
+		ProgramName: "data_feeds_cache",
+		FeedAdmins:  req.FeedAdmins,
+	}
+
+	deps := operation.Deps{
+		Datastore: env.DataStore,
+		Env:       env,
+		Chain:     ch,
+	}
+
+	deploySeqReport, err := operations.ExecuteSequence(env.OperationsBundle, seq.DeployCacheSeq, deps, deploySeqInput)
+	if err != nil {
+		return out, err
+	}
+
+	// Save programID
+	err = out.DataStore.Addresses().Add(
+		datastore.AddressRef{
+			Address:       deploySeqReport.Output.ProgramID.String(),
+			ChainSelector: req.ChainSel,
+			Type:          CacheContract,
+			Version:       version,
+			Qualifier:     req.Qualifier,
+			Labels:        req.LabelSet,
+		},
+	)
+	if err != nil {
+		return out, err
+	}
+	// Save StateID
+	err = out.DataStore.Addresses().Add(
+		datastore.AddressRef{
+			Address:       deploySeqReport.Output.State.String(),
```

Comment:
what's the difference between StateID and ProgramID ? If we save both in datastore with the exact qualifier and labels, it would be very hard to query the correct one later on